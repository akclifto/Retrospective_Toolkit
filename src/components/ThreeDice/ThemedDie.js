import React, { useEffect } from "react";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { uniqueImageSet } from "../Dice/Dice";

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    dicePos,
    rerollAllToggle,
    rerollValue,
    rerollDieToggle,
    imageSet,
    setImages,
  } = props;
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    rotation: [
      Math.random() * Math.PI,
      Math.random() * Math.PI,
      Math.random() * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  useEffect(() => {
    api.position.set(dicePos[0], dicePos[1], dicePos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    setImages(uniqueImageSet);
  }, [
    api.angularVelocity,
    api.position,
    api.velocity,
    dicePos,
    rerollAllToggle,
    rerollValue,
    setImages,
  ]);
  return (
    <mesh
      onClick={() => {
        rerollDieToggle(!rerollValue);
      }}
      ref={mesh}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  dicePos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  setImages: PropTypes.func.isRequired,
  rerollAllToggle: PropTypes.bool.isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
};

export default ThemedDie;
