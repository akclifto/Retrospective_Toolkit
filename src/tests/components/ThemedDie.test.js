import React, { Suspense } from "react";
import { shallow } from "enzyme";
// import { useBox } from "@react-three/cannon";

import ThemedDie from "../../components/ThreeDice/ThemedDie";

/*
jest.mock("@react-three/cannon", () => ({
  useBox: () => ({
    mesh: {
      mass: 300,
      inertia: 13,
      rotation: [
        Math.random() * Math.PI,
        Math.random() * Math.PI,
        Math.random() * Math.PI,
      ],
      linearDamping: 0.5,
      angularDamping: 0.1,
      material: { restitution: 0.3 },
    },
    api: {
      position: [0, 0, 0],
      rotation: [0, 0, 0],
      velocity: [0, 0, 0],
      angularVelocity: [0, 0, 0],
      linearFactor: [0, 0, 0],
      angularFactor: [0, 0, 0],
      applyForce: jest.fn(),
      applyImpulse: jest.fn(),
      applyLocalForce: jest.fn(),
      applyLocalImpulse: jest.fn(),
    },
  }),
}));
*/
describe("ThreeDice/ThemedDie Testing", () => {
  it("should do a snapshot test on ThemedDie", async () => {
    const wrapper = shallow(
      <Suspense fallback={<div>Loading...</div>}>
        <ThemedDie
          dicePos={[-13, 5, 6]}
          rerollValue={false}
          rerollAllToggle={false}
          rerollDieToggle={() => {}}
          imageSet={[""]}
          setImages={() => {}}
          testID="testDie"
        />
      </Suspense>
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
