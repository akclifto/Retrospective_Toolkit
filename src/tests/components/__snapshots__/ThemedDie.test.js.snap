// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeDice/ThemedDie Testing should do a snapshot test on ThemedDie 1`] = `
<Suspense
  fallback={
    <div>
      Loading...
    </div>
  }
>
  <ThemedDie
    dicePos={
      Array [
        -13,
        5,
        6,
      ]
    }
    imageSet={
      Array [
        "",
      ]
    }
    rerollAllToggle={false}
    rerollDieToggle={[Function]}
    rerollValue={false}
    setImages={[Function]}
    testID="testDie"
  />
</Suspense>
`;
